File: app/src/main/java/com/example/hitchapp/adapters/MessagesAdapter.java
package com.example.hitchapp.adapters;

import android.content.Context;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.example.hitchapp.R;
import com.example.hitchapp.models.Message;
import com.example.hitchapp.models.Post;
import com.example.hitchapp.models.User;
import com.parse.ParseUser;

import java.math.BigInteger;
import java.security.MessageDigest;
import java.util.List;

public class MessagesAdapter extends RecyclerView.Adapter<MessagesAdapter.ViewHolder> {
    private List<Message> mMessages;
    private Context mContext;
    private String mUserId;

    public MessagesAdapter(Context context, String userId, List<Message> messages) {
        mMessages = messages;
        this.mUserId = userId;
        mContext = context;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        Context context = parent.getContext();
        LayoutInflater inflater = LayoutInflater.from(context);
        View contactView = inflater.inflate(R.layout.item_message, parent, false);

        ViewHolder viewHolder = new ViewHolder(contactView);
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        Message message = mMessages.get(position);
        // Gets the person who's logged in
        User user = (User) ParseUser.getCurrentUser();
        final boolean isMe = message.getAuthor().getObjectId().equals(user.getObjectId());

        if (isMe) {
            holder.imageMe.setVisibility(View.VISIBLE);
            holder.imageOther.setVisibility(View.GONE);
            holder.body.setGravity(Gravity.CENTER_VERTICAL | Gravity.RIGHT);
        } else {
            holder.imageOther.setVisibility(View.VISIBLE);
            holder.imageMe.setVisibility(View.GONE);
            holder.body.setGravity(Gravity.CENTER_VERTICAL | Gravity.LEFT);
        }

        final ImageView profileView = isMe ? holder.imageMe : holder.imageOther;
        Glide.with(mContext).load(getProfileUrl(message.getAuthor().getObjectId())).into(profileView);
        holder.body.setText(message.getContent());
    }

    // Create a gravatar image based on the hash value obtained from userId
    private static String getProfileUrl(final String userId) {
        String hex = "";
        try {
            final MessageDigest digest = MessageDigest.getInstance("MD5");
            final byte[] hash = digest.digest(userId.getBytes());
            final BigInteger bigInt = new BigInteger(hash);
            hex = bigInt.abs().toString(16);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return "https://www.gravatar.com/avatar/" + hex + "?d=identicon";
    }

    @Override
    public int getItemCount() {
        return mMessages.size();
    }

    // Clean all elements of the recycler
    public void clear() {
        mMessages.clear();
        notifyDataSetChanged();
    }

    // Add a list of items -- change to type used
    public void addAll(List<Message> list) {
        mMessages.addAll(list);
        notifyDataSetChanged();
    }

    public void setAll(List<Message> list){
        clear();
        addAll(list);
        notifyDataSetChanged();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        ImageView imageOther;
        ImageView imageMe;
        TextView body;

        public ViewHolder(View itemView) {
            super(itemView);
            imageOther = (ImageView)itemView.findViewById(R.id.ivProfileOther);
            imageMe = (ImageView)itemView.findViewById(R.id.ivProfileMe);
            body = (TextView)itemView.findViewById(R.id.tvBody);
        }
    }
}

File: app/src/main/java/com/example/hitchapp/adapters/MessagesAdapter.kt
package com.example.hitchapp.adapters

import android.content.Context
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.example.hitchapp.R
import com.example.hitchapp.models.Message
import com.example.hitchapp.models.Request
import com.example.hitchapp.models.User
import com.parse.ParseException
import com.parse.ParseFile
import com.parse.ParseUser
import java.math.BigInteger
import java.security.MessageDigest

class MessagesAdapter(private val mContext: Context, private val mUserId: String, private val mMessages: MutableList<Message>) : RecyclerView.Adapter<MessagesAdapter.ViewHolder>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val context = parent.context
        val inflater = LayoutInflater.from(context)
        val contactView = inflater.inflate(R.layout.item_message, parent, false)
        return ViewHolder(contactView)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val message = mMessages[position]
        // Gets the person who's logged in
        val user = ParseUser.getCurrentUser() as User
        val isMe = message.author!!.objectId == user.objectId
        if (isMe) {
            holder.imageMe.visibility = View.VISIBLE
            holder.imageOther.visibility = View.GONE
            holder.body.gravity = Gravity.CENTER_VERTICAL or Gravity.RIGHT
        } else {
            holder.imageOther.visibility = View.VISIBLE
            holder.imageMe.visibility = View.GONE
            holder.body.gravity = Gravity.CENTER_VERTICAL or Gravity.LEFT
        }
        val profileView = if (isMe) holder.imageMe else holder.imageOther
        var profile: ParseFile? = null
        try {
            val userPic = message.author!!.fetch() as User
            profile = userPic.profilePicture
        } catch (e: ParseException) {
            e.printStackTrace()
        }
        if (profile != null) {
            Glide.with(mContext)
                    .load(profile.url)
                    .fitCenter()
                    .circleCrop()
                    .into(profileView)
        }
        //Glide.with(mContext).load(getProfileUrl(message.getAuthor().getObjectId())).into(profileView);
        holder.body.text = message.content
    }

    override fun getItemCount(): Int {
        return mMessages.size
    }

    // Clean all elements of the recycler
    fun clear() {
        mMessages.clear()
        notifyDataSetChanged()
    }

    // Add a list of items -- change to type used
    fun addAll(list: Collection<Message?>) {
        mMessages.addAll(list as Collection<Message>)
        notifyDataSetChanged()
    }

    fun setAll(list: List<Message?>?) {
        clear()
        list?.let { addAll(it) }
        notifyDataSetChanged()
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var imageOther: ImageView = itemView.findViewById<View>(R.id.ivProfileOther) as ImageView
        var imageMe: ImageView = itemView.findViewById<View>(R.id.ivProfileMe) as ImageView
        var body: TextView = itemView.findViewById<View>(R.id.tvBody) as TextView

    }

    companion object {
        // Create a gravatar image based on the hash value obtained from userId
        private fun getProfileUrl(userId: String): String {
            var hex = ""
            try {
                val digest = MessageDigest.getInstance("MD5")
                val hash = digest.digest(userId.toByteArray())
                val bigInt = BigInteger(hash)
                hex = bigInt.abs().toString(16)
            } catch (e: Exception) {
                e.printStackTrace()
            }
            return "https://www.gravatar.com/avatar/$hex?d=identicon"
        }
    }

}